src/openms/source/CHEMISTRY/RibonucleotideDB.cpp

```diff
@@ -66,6 +87,209 @@ namespace OpenMS
     }
   }
 
+  // All valid JSON ribonucleotides must at minimum have elements defining name, short_name, reference_moiety, and formula
+  bool entryIsWellFormed_(const nlohmann::json::value_type& entry)
+  {
+    if (entry.find("name") == entry.cend())
+    {
+      String msg = "\"name\" entry missing for ribonucleotide";
+      throw Exception::MissingInformation(__FILE__, __LINE__,
+                                              OPENMS_PRETTY_FUNCTION, msg);
+    }
+    if (entry.find("short_name") == entry.cend())
+    {
+      String msg = "\"short_name\" entry missing for ribonucleotide";
+      throw Exception::MissingInformation(__FILE__, __LINE__,
+                                              OPENMS_PRETTY_FUNCTION, msg);
+    }
+    if (entry.find("reference_moiety") == entry.cend())
+    {
+      String msg = "\"reference_moiety\" entry missing for ribonucleotide";
+      throw Exception::MissingInformation(__FILE__, __LINE__,
+                                              OPENMS_PRETTY_FUNCTION, msg);
+    }
+    if (entry.find("formula") == entry.cend())
+    {
+      String msg = "\"formula\" entry missing for ribonucleotide";
+      throw Exception::MissingInformation(__FILE__, __LINE__,
+                                              OPENMS_PRETTY_FUNCTION, msg);
+    }
+    return true;
+  }
+  
+  EmpiricalFormula getBaseLossFormula_(const nlohmann::json::value_type& entry)
+  {
+    String code = entry.at("short_name");
+    // If we have an explicitly defined baseloss_formula
+    if (auto e = entry.find("baseloss_formula"); e != entry.cend() && !e->is_null())
+    {
+      return EmpiricalFormula(entry.at("baseloss_formula"));
+    }
+    //TODO: Calculate base loss formula from SMILES
+    else // If we don't have a defined baseloss_formula calculate it from our shortCode
+    {
+      if (code.hasPrefix('d')) // handle deoxyribose, possibly with methyl mod
+      {
+        return EmpiricalFormula("C5H10O4");
+      }
+      else if (code.hasSuffix('m')) // mod. attached to the ribose, not base
+      {
+        return EmpiricalFormula("C6H12O5");
+      }
+      else if (code.hasSuffix("m*")) // check if we have both a sulfer and a 2'-O methyl
+      {
+        return EmpiricalFormula("C6H12O5");
+      }
+      else if (code.hasSuffix("Ar(p)") ||  code.hasSuffix("Gr(p)"))
+      {
+        return EmpiricalFormula("C10H19O21P");
+      }
+      else
+      {
+        return EmpiricalFormula("C5H10O5");
+      }
+    }
+  }
+
+  ParsedEntry_ parseEntry_(const nlohmann::json::value_type& entry)
+  {
+    ParsedEntry_ parsed;
+    Ribonucleotide* ribo = new Ribonucleotide();
```

Comment:
this should really be a `unique_ptr<Ribonucleotide>`.
The exceptions below will cause memory leaks otherwise.

`ribonucleotides_` should also contain unique_ptr's. This avoids having a custom Dtor etc...

Reply:
I've handled the exceptions here (freeing ribo before throwing.). I'll open an issue for the larger suggestion.